README.md
# RunControl

## Introduction

This is a Hades modpack designed to replicate, in detail, the events of an already-played run. It is also versatile enough to be used for testing, proof-of-concepting, and racing. The player can replicate the decisions originally made, or make different ones, and the modpack will respond correctly to the changing conditions this causes.

## Installation and Usage

To install the modpack, download your desired version from the releases page, extract it into your Hades\Content folder, and run modimporter.exe. If you have any problems, you can ask in the #modification-station channel in the Hades Speedrunning Discord, or #install-help in the Hades Modding Discord.

Once the modpack is installed, enter the courtyard in-game and press the boon menu button. If you've installed correctly, you'll see a RunControl button, where you can select a run and configure the mods.

## Creating a run

Each run is one .lua file, stored in RunControl\\Runs\\. An explanation of the folder and file structure can be found in the [HOWTO](https://github.com/Hades-Speedrunning/RunControl/blob/main/HOWTO.md) file.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- Support for high heat features- Benefits Package, Jury Summons, and Approval Process.
- Support for Anvils.
- Support for Hades' additional spawns.

## [1.2.0] - 2023-12-05

Third full release! This version contains some minor fixes compared to RC1.

### Added

- There's now an AlwaysEligible field in the theseusEncounter data type.
- Updated HOWTO.md to explain the new 1.2.0 features.

### Fixed

- Wells with only one forced item can no longer show duplicates of that item.

*...plus changes from 1.2.0-RC1*

## [1.2.0-RC1] - 2023-12-01

This update adds several features for better configurability and ease of run creation, as well as several bug fixes and under-the-hood refactors. If no critical bugs are found, this will be released as 1.2.0.

### Added

- Option for special rooms (fountain, story, shop, midboss) to only appear if forced.
- Option for rarities to only be forced when eligible (e.g., Common boons won't appear when you have a Yarn).
- Option for gold pots to only be forced up to the maximum amount a room allows.
- Encounters can now specify enemy types and wave counts, rather than having to include exact spawn counts.
- Runs can now specify which God Theseus should summon.
- RunControl is now compatible with extensions like StartingBuildMod.
- A specific God can now be forced in chamber 1, just like in other chambers.

### Changed

- The wells and pools in midbiomes will now appear even if a run does not force them. (They can still be forced not to appear.)
- Encounters/Bosses have now been split into a new menu page, separate from Rooms/Rewards.
- Sell wells will now infer the value of a boon if none is specified, instead of defaulting to 0.

### Fixed

- Wells and shops will no longer fail to force items if their data has gaps (e.g. if you specify items 1 and 3 but not 2).
- You can now force Chaos gates not to appear, even if you have a Light of Ixion.
- Rare Crop will no longer fail to appear when forced.
- Fixed some errors in ExampleRun.lua.
- The default rarity setting will no longer cause hammers to have rarity.

## [1.1.0] - 2023-10-04

This is a smaller update, which doesn't add many new features to the runs themselves, but adds a little more polish to 1.0.0 (plus some new options).

### Added

- Interface/QOL settings page.
- Option to always show chamber number in the top-right.
- Option to replace the modpack version text with the name of the current run.

### Changed

- Chaos gates are no longer counted in exitDoors.

### Fixed

- The D_Mini09 chamber in Styx will no longer have random encounters.
- Wells will no longer occasionally have an incorrect item after rerolling.
- Corrected a room orientation error in the 6:07 Talos run.
- Bosses and Styx midbosses will no longer need to be AlwaysEligible in order to appear, and will not appear on non-midboss Styx wings.
- The ExampleRun file no longer has a missing value in IndexedBy.
- Indexing Gods by appearanceNum will now work correctly in Styx shop.
- Forcing an empty slot as the only Boon from a God will no longer cause a crash.
- Getting a Light of Ixion the original player did not get will no longer change the possible doors in a run.

## [1.0.0] - 2023-09-29

First full release!

### Added

- Added Chaos boon documentation to HOWTO.md.
- Added a special thanks section to the Introduction screen.

### Changed

- Updated the Introduction text slightly.

## [1.0.0-RC3] - 2023-09-28

### Fixed

- Chamber 1 encounter will no longer be random.
- Rewards can no longer be offered on story rooms and shops.

## [1.0.0-RC2] - 2023-09-28

### Fixed

- RewardControl gods will no longer take precedence over those forced with a keepsake.
- EncounterControl will no longer try to force waves where it shouldn't, such as in midboss and non-combat rooms.

## [1.0.0-RC1] - 2023-09-27

### Added

- Implemented the 5:04 current WR by DrkFrst and the 5:57 former WR by Lili.
- Expanded the README and added a HOWTO file, explaining the run format.

### Changed

- BoonControl's InferReplaces config option is now off by default.

### Fixed

- Using multiple gods in the same room will now work again.

## [1.0.0-beta.7] - 2023-09-27

### Fixed

- Gods will now only be offered if they are in your pool.
- Replaces will no longer always be eligible.

## [1.0.0-beta.6] - 2023-09-26

### Added

- Implemented the 5:49 Achilles WR by 185.

### Fixed

- BoonControl's FillWithEligible config option will now work on Chaos.

## [1.0.0-beta.5] - 2023-09-24

### Added

- Implemented the 6:29 former Demeter WR by Ananke.

### Changed

- Lists that don't, at any point, specify a dataType will now throw out an error without successfully returning anything. This prevents unintended behaviour or crashes that not specifying could potentially cause.

### Fixed

- Elite rooms with Fated Key rewards will now work correctly.

## [1.0.0-beta.4] - 2023-09-22

### Added

- Implemented the 5:16 Hera WR by Croven.

### Changed

- Runs to be implemented in the further future have been removed from the menu for the time being.

### Fixed

- The game will no longer crash when you buy a random bag without specified contents.

## [1.0.0-beta.3] - 2023-09-19

### Added


RunControl/RunControl.lua
--[[
    RunControl
    Author: SleepSoul (Discord: sleepsoul)
    Dependencies: ModUtil, RCLib
    Central mod used to set the run data for all mods in the RunControl modpack.
]]
ModUtil.Mod.Register( "RunControl" )

local config = {
    SelectedRun = "None",
    ShowChamberNumber = true,
    ShowRunName = true,
}
RunControl.config = config

RunControl.ModpackVersion = "1.2.0"

RunControl.CurrentRunData = {}
RunControl.Runs = {}
RunControl.SortedRuns = {
    "Lili557Eris",
    "Ananke552Eris",
    "185549Achilles",
    "Croven532Beowulf",
    "ocaml526Rama",
    "Croven516Hera",
    "DrkFrst504Beowulf",

    "BIGP524Beowulf",
    "185455Achilles",
    "Croven449Hera",
    "BIGP448Beowulf",
    "Croven447Hera",
    "Ocke445Beowulf",
    "Croven439Hera",
    "Vorime430Beowulf",

    "SleepSoul607Talos",
    "CherryDad541Gilgamesh",

    "Ananke629Demeter",
}

RunControl.ModIndex = {
    "BoonControl",
    "BossControl",
    "EncounterControl",
    "RewardControl",
    "RoomControl",
    "SellControl",
    "ShopControl",
}

function RunControl.DisplayVersion()
    local textConfigTable = {
        color = Color.DimGray,
        x_pos = 1908,
        y_pos = 27,
    }
    local versionString = "RunControl " .. RunControl.ModpackVersion
    local runName = ModUtil.IndexArray.Get( RunControl.Runs, { RunControl.config.SelectedRun, "Metadata", "Name" } )

    if RunControl.config.ShowRunName and runName and CurrentDeathAreaRoom == nil then
        versionString = runName
    end

    PrintUtil.createOverlayLine( "ModpackVersion", versionString, textConfigTable )
end

function RunControl.UpdateRunData()
    RunControl.CurrentRunData = RunControl.Runs[RunControl.config.SelectedRun] or {}
    for _, modName in ipairs( RunControl.ModIndex ) do
        if ModUtil.Path.Get( modName ) then
            ModUtil.Path.Set( modName .. ".CurrentRunData", RunControl.CurrentRunData )
        end
    end
end

function RunControl.SaveConfig()
    RunControl.Data.config = {}
    for k, v in pairs( RunControl.config ) do
        RunControl.Data.config[k] = v
    end

    for _, modName in ipairs( RunControl.ModIndex ) do
        local modConfig = ModUtil.Path.Get( modName .. ".config" ) or {}

        if not IsEmpty( modConfig ) and modConfig.SaveToRunControl ~= false then -- We don't want to create a Data.config table to save into unless there's something to save
            ModUtil.IndexArray.Set( _G, { modName, "Data", "config" }, {} ) -- Create (or empty) the Data.config table for this mod
            for k, v in pairs( modConfig ) do
                ModUtil.IndexArray.Set( _G, { modName, "Data", "config", k }, v ) -- Copy from <modName>.config (which is not saved) to <modName>.Data.config (which is)
            end
        end
    end
end

function RunControl.LoadConfig()
    if not IsEmpty( ModUtil.Path.Get( "RunControl.Data.config" ) ) then
        for k, v in pairs( RunControl.Data.config ) do
            RunControl.config[k] = v
        end
    end

    for _, modName in ipairs( RunControl.ModIndex ) do
        local savedModConfig = ModUtil.Path.Get( modName .. ".Data.config" ) or {}
        local useSaveData = ModUtil.Path.Get( modName .. ".config.SaveToRunControl" ) -- Allows mods to opt out of the save data system

        if not IsEmpty( savedModConfig ) and useSaveData ~= false then
            for k, v in pairs( savedModConfig ) do
                ModUtil.IndexArray.Set( _G, { modName, "config", k }, v ) -- Copy from <modName>.Data.config (saved data) to <modName>.config (actively used)
            end
        end
    end
end

ModUtil.LoadOnce( function()
    UIData.CurrentRunDepth.X = 1780 -- Change positioning of chamber number indicator to fit better with our UI
    UIData.CurrentRunDepth.Y = 60
    UIData.CurrentRunDepth.TextFormat.ShadowColor = { 0, 0, 0, 0 }
    RunControl.LoadConfig()
    RunControl.UpdateRunData()
end )

OnAnyLoad{ function()
    RunControl.DisplayVersion()
end }

ModUtil.Path.Wrap( "StartRoom", function( baseFunc, currentRun, currentRoom )
    if RunControl.config.ShowChamberNumber then
        ShowDepthCounter()
    end
    baseFunc( currentRun, currentRoom )
end, RunControl )

ModUtil.Path.Wrap("ShowCombatUI", function( baseFunc, flag )
    if RunControl.config.ShowChamberNumber then
        ShowDepthCounter()
    end
    baseFunc( flag )
end, RunControl)

ModUtil.Path.Wrap( "RunClearMessagePresentation", function( baseFunc, screen, message, args )
    message = message or {}
    message.Name = ModUtil.IndexArray.Get( RunControl.Runs, { RunControl.config.SelectedRun, "Metadata", "Name" } ) or "RunControl"
    baseFunc( screen, message, args )
end, RunControl )

ModUtil.Path.Override( "ShowRunClearScreen", function()

	PlaySound({ Name = "/Leftovers/Menu Sounds/AscensionConfirm" })

	RecordRunCleared()

	thread( PlayVoiceLines, HeroVoiceLines.RunClearedVoiceLines )

	ScreenAnchors.RunClear = { Components = {} }
	local screen = ScreenAnchors.RunClear
	screen.Name = "RunClear"

	if IsScreenOpen( screen.Name ) then
		return
	end
	OnScreenOpened({ Flag = screen.Name, PersistCombatUI = true })
	FreezePlayerUnit()
	EnableShopGamepadCursor()

	ToggleControl({ Names = { "AdvancedTooltip" }, Enabled = false })
	thread( ShowAdvancedTooltip, { DontDuckAudio = true, DisableTooltips = true, HideCloseButton = true, AutoPin = true, } )

	PlaySound({ Name = "/SFX/Menu Sounds/DialoguePanelIn" })

	CreateGroup({ Name = "Combat_Menu_Overlay2" })
	InsertGroupInFront({ Name = "Combat_Menu_Overlay2", DestinationName = "Combat_Menu_Overlay" })

	local components = screen.Components

	components.Blackout = CreateScreenComponent({ Name = "rectangle01", Group = "Combat_UI_Backing", X = ScreenCenterX, Y = ScreenCenterY })
	SetScale({ Id = components.Blackout.Id, Fraction = 10 })
	SetColor({ Id = components.Blackout.Id, Color = Color.Black })
	SetAlpha({ Id = components.Blackout.Id, Fraction = 0 })
	SetAlpha({ Id = components.Blackout.Id, Fraction = 0.6, Duration = 0.5 })

	components.ShopBackground = CreateScreenComponent({ Name = "EndPanelBox", Group = "Combat_Menu_Overlay2", X = ScreenCenterX + 637, Y = ScreenCenterY - 30 })
	for cosmeticName, status in pairs( GameState.Cosmetics ) do
		local cosmeticData = ConditionalItemData[cosmeticName]
		if cosmeticData ~= nil and cosmeticData.RunClearScreenBacking ~= nil then
			SetAnimation({ DestinationId = components.ShopBackground.Id, Name = cosmeticData.RunClearScreenBacking })
			break
		end
	end
	components.CloseButton = CreateScreenComponent({ Name = "ButtonClose", Group = "Combat_Menu_TraitTray", Scale = 0.7 })
	Attach({ Id = components.CloseButton.Id, DestinationId = components.ShopBackground.Id, OffsetX = 3, OffsetY = 480 })
	components.CloseButton.OnPressedFunctionName = "CloseRunClearScreen"
	components.CloseButton.ControlHotkey = "Cancel"

	-- Title
	CreateTextBox(MergeTables({ Id = components.ShopBackground.Id,
		Text = "RunClearScreen_Title",
		FontSize = 32,
		Font = "SpectralSCLightTitling",
		OffsetX = -4, OffsetY = -370,
		Color = Color.White,
		ShadowBlur = 0, ShadowColor = {0,0,0,1}, ShadowOffset={0, 3}, Justification = "Center",
		}, LocalizationData.RunClearScreen.TitleText))

	local messageOffsetX = -255
	local statOffsetX = 100

	local recordTime = GetFastestRunClearTime( CurrentRun )
	local prevRecordShrinePoints = GetHighestShrinePointRunClear()

	local offsetY = -325

	local lineSpacingLarge = 55
	local lineSpacingSmall = 35

	local mainFontSize = 19
	local titleColor = Color.White
	local dataColor = {0.702, 0.620, 0.345, 1.0}
	local newRecordColor = {1.000, 0.894, 0.231, 1.0}

    local originalTime = ModUtil.IndexArray.Get( RunControl.Runs, { RunControl.config.SelectedRun, "Metadata", "OriginalTime" } )

	-- ClearTime
	offsetY = offsetY + lineSpacingLarge
	CreateTextBox({ Id = components.ShopBackground.Id,
		Text = "RunClearScreen_ClearTime",
		FontSize = mainFontSize,
		OffsetX = messageOffsetX, OffsetY = offsetY,
		Color = titleColor,
		Font = "AlegreyaSansSCRegular",
		ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
		Justification = "Left" })
	CreateTextBox({ Id = components.ShopBackground.Id,
		Text = GetTimerString( CurrentRun.GameplayTime, 2 ),
		FontSize = mainFontSize,
		OffsetX = statOffsetX, OffsetY = offsetY,
		Color = dataColor,
		Font = "AlegreyaSansSCExtraBold",
		ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
		Justification = "Right" })
	if not originalTime and CurrentRun.GameplayTime <= recordTime then
		wait(0.03)
		CreateTextBox(MergeTables({ Id = components.ShopBackground.Id,
			Text = "RunClearScreen_NewRecord",
			FontSize = mainFontSize,
			OffsetX = statOffsetX + 20, OffsetY = offsetY,
			Color = newRecordColor,
			Font = "AlegreyaSansSCExtraBold",
			ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
			Justification = "Left" }, LocalizationData.RunClearScreen.NewRecordText))
	end

	wait(0.05)

	offsetY = offsetY + lineSpacingSmall
	-- Record Clear Time/Original Time
    if originalTime then
        CreateTextBox({ Id = components.ShopBackground.Id,
            Text = "Original Time:",
            FontSize = mainFontSize,
            OffsetX = messageOffsetX, OffsetY = offsetY,
            Color = titleColor,
            Font = "AlegreyaSansSCRegular",
            ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
            Justification = "Left" })
        CreateTextBox({ Id = components.ShopBackground.Id,
            Text = originalTime,
            FontSize = mainFontSize,
            OffsetX = statOffsetX, OffsetY = offsetY,
            Color = dataColor,
            Font = "AlegreyaSansSCExtraBold",
            ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
            Justification = "Right" })
    else
        CreateTextBox({ Id = components.ShopBackground.Id,
            Text = "RunClearScreen_ClearTimeRecord",
            FontSize = mainFontSize,
            OffsetX = messageOffsetX, OffsetY = offsetY,
            Color = titleColor,
            Font = "AlegreyaSansSCRegular",
            ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
            Justification = "Left" })
        CreateTextBox({ Id = components.ShopBackground.Id,
            Text = GetTimerString( recordTime, 2 ),
            FontSize = mainFontSize,
            OffsetX = statOffsetX, OffsetY = offsetY,
            Color = dataColor,
            Font = "AlegreyaSansSCExtraBold",
            ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
            Justification = "Right" })
    end

    local originalHeat = ModUtil.IndexArray.Get( RunControl.Runs, { RunControl.config.SelectedRun, "Metadata", "OriginalHeat" } )

	-- ShrinePoints
	offsetY = offsetY + lineSpacingLarge
	CreateTextBox({ Id = components.ShopBackground.Id,
		Text = "RunClearScreen_ShrinePoints",
		FontSize = mainFontSize,
		OffsetX = messageOffsetX, OffsetY = offsetY,
		Color = titleColor,
		Font = "AlegreyaSansSCRegular",
		ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
		Justification = "Left" })
	CreateTextBox({ Id = components.ShopBackground.Id,
		Text = CurrentRun.ShrinePointsCache,
		FontSize = mainFontSize,
		OffsetX = statOffsetX, OffsetY = offsetY,
		Color = dataColor,
		Font = "AlegreyaSansSCExtraBold",
		ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
		Justification = "Right" })
	if not originalHeat and CurrentRun.ShrinePointsCache > prevRecordShrinePoints then
		wait(0.03)
		CreateTextBox(MergeTables({ Id = components.ShopBackground.Id,
			Text = "RunClearScreen_NewRecord",
			FontSize = mainFontSize,
			OffsetX = statOffsetX + 20, OffsetY = offsetY,
			Color = newRecordColor,
			Font = "AlegreyaSansSCExtraBold",
			ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
			Justification = "Left" }, LocalizationData.RunClearScreen.NewRecordText))
	end

	wait(0.05)

	-- Record ShrinePoints
	offsetY = offsetY + lineSpacingSmall
    if originalHeat then
        CreateTextBox({ Id = components.ShopBackground.Id,
            Text = "{!Icons.ShrinePointSmall_Active} Original:",
            FontSize = mainFontSize,
            OffsetX = messageOffsetX, OffsetY = offsetY,
            Color = titleColor,
            Font = "AlegreyaSansSCRegular",
            ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
            Justification = "Left" })
        CreateTextBox({ Id = components.ShopBackground.Id,
            Text = originalHeat,
            FontSize = mainFontSize,
            OffsetX = statOffsetX, OffsetY = offsetY,
            Color = dataColor,
            Font = "AlegreyaSansSCExtraBold",
            ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
            Justification = "Right" })
    else
        CreateTextBox({ Id = components.ShopBackground.Id,
            Text = "RunClearScreen_ShrinePointsRecord",
            FontSize = mainFontSize,
            OffsetX = messageOffsetX, OffsetY = offsetY,
            Color = titleColor,
            Font = "AlegreyaSansSCRegular",
            ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
            Justification = "Left" })
        CreateTextBox({ Id = components.ShopBackground.Id,
            Text = math.max( CurrentRun.ShrinePointsCache, prevRecordShrinePoints ),
            FontSize = mainFontSize,
            OffsetX = statOffsetX, OffsetY = offsetY,
            Color = dataColor,
            Font = "AlegreyaSansSCExtraBold",
            ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
            Justification = "Right" })
    end

	wait(0.03)

	-- Weapon Clears
	offsetY = offsetY + 90

	local columnHeaders =
	{
		{ OffsetX = 0, Text = "RunClearScreen_Header_Weapon", Justification = "Left", },
		{ OffsetX = 254, Text = "RunClearScreen_Header_Clears", Justification = "Right", },
		{ OffsetX = 384, Text = "RunClearScreen_Header_RecordClearTime", Justification = "Right", },
		{ OffsetX = 496, Text = "RunClearScreen_Header_RecordShrinePoints", Justification = "Right", },
	}

	for k, header in ipairs( columnHeaders ) do
		CreateTextBox(MergeTables({ Id = components.ShopBackground.Id,
				Text = header.Text,
				FontSize = mainFontSize,
				OffsetX = messageOffsetX + header.OffsetX, OffsetY = offsetY,
				Color = Color.White,
				Font = "AlegreyaSansSCExtraBold",
				ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
				Justification = header.Justification }, LocalizationData.RunClearScreen.ColumnHeader))
	end

	offsetY = offsetY + lineSpacingLarge
	wait(0.05)

	for k, weaponName in ipairs( WeaponSets.HeroMeleeWeapons ) do

		local weaponNameFont = "AlegreyaSansSCRegular"
		if CurrentRun.Hero.Weapons[weaponName] then
			-- Currently used weapon
			weaponNameFont = "AlegreyaSansSCExtraBold"
		end

		local weaponText = weaponName
		if not IsWeaponUnlocked( weaponName ) then
			weaponText = "MysteryUpgrade"
		end

		CreateTextBox(MergeTables({ Id = components.ShopBackground.Id,
			Text = weaponText,
			FontSize = mainFontSize,
			OffsetX = messageOffsetX + columnHeaders[1].OffsetX, OffsetY = offsetY,
			Color = {0.569, 0.557, 0.533, 1.0},
			Font = weaponNameFont,
			ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
			Justification = columnHeaders[1].Justification }, LocalizationData.RunClearScreen.WeaponText))

		local weaponClears = GetNumRunsClearedWithWeapon( weaponName )
		CreateTextBox(MergeTables({ Id = components.ShopBackground.Id,
			Text = weaponClears,
			FontSize = mainFontSize,
			OffsetX = messageOffsetX + columnHeaders[2].OffsetX, OffsetY = offsetY,
			Color = dataColor,
			Font = "AlegreyaSansSCExtraBold",
			ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
			Justification = columnHeaders[2].Justification }, LocalizationData.RunClearScreen.WeaponText))

		local weaponTime = GetFastestRunClearTimeWithWeapon( CurrentRun, weaponName )
		if weaponTime ~= nil then
			CreateTextBox(MergeTables({ Id = components.ShopBackground.Id,
				Text = GetTimerString( weaponTime, 2 ),
				FontSize = mainFontSize,
				OffsetX = messageOffsetX + columnHeaders[3].OffsetX, OffsetY = offsetY,
				Color = dataColor,
				Font = "AlegreyaSansSCExtraBold",
				ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
				Justification = columnHeaders[3].Justification }, LocalizationData.RunClearScreen.WeaponText))
		end

		local weaponShrinePoints = GetHighestShrinePointRunClearWithWeapon( CurrentRun, weaponName )
		if weaponShrinePoints > 0 then
			CreateTextBox(MergeTables({ Id = components.ShopBackground.Id,
				Text = weaponShrinePoints,
				FontSize = mainFontSize,
				OffsetX = messageOffsetX + columnHeaders[4].OffsetX - 10, OffsetY = offsetY,
				Color = dataColor,
				Font = "AlegreyaSansSCExtraBold",
				ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
				Justification = columnHeaders[4].Justification }, LocalizationData.RunClearScreen.WeaponText))
			end

		offsetY = offsetY + ( lineSpacingSmall * 1.1 )
		wait(0.03)

	end

	wait(0.05)

    local selectedRun = ModUtil.Path.Get( "RunControl.config.SelectedRun" ) or "None"
    local clearsOnThisSeed = 0
    if selectedRun ~= "None" then
        clearsOnThisSeed = ( ModUtil.IndexArray.Get( RunControl.Data, { "RunHistory", selectedRun, "Clears" } ) or 0 ) + 1
        ModUtil.IndexArray.Set( RunControl.Data, { "RunHistory", selectedRun, "Clears" }, clearsOnThisSeed )
    end

	-- Total Clears
	offsetY = offsetY + 67
	CreateTextBox({ Id = components.ShopBackground.Id,
		Text = "RunClearScreen_TotalClears",
		FontSize = mainFontSize,
		OffsetX = messageOffsetX, OffsetY = offsetY,
		Color = titleColor,
		Font = "AlegreyaSansSCRegular",
		ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
		Justification = "Left" })
	CreateTextBox({ Id = components.ShopBackground.Id,
		Text = GameState.TimesCleared,
		FontSize = mainFontSize,
		OffsetX = statOffsetX, OffsetY = offsetY,
		Color = dataColor,
		Font = "AlegreyaSansSCExtraBold",
		ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
		Justification = "Right" })

	-- Clears on this seed
	offsetY = offsetY + lineSpacingSmall
    if selectedRun == "None" then
        CreateTextBox({ Id = components.ShopBackground.Id,
            Text = "No Run Selected",
            FontSize = mainFontSize,
            OffsetX = messageOffsetX, OffsetY = offsetY,
            Color = titleColor,
            Font = "AlegreyaSansSCRegular",
            ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
            Justification = "Left"
        })
    else
        CreateTextBox({ Id = components.ShopBackground.Id,
            Text = "Run Clears:",
            FontSize = mainFontSize,
            OffsetX = messageOffsetX, OffsetY = offsetY,
            Color = titleColor,
            Font = "AlegreyaSansSCRegular",
            ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
            Justification = "Left"
        })
        CreateTextBox({ Id = components.ShopBackground.Id,
            Text = clearsOnThisSeed,
            FontSize = mainFontSize,
            OffsetX = statOffsetX, OffsetY = offsetY,
            Color = dataColor,
            Font = "AlegreyaSansSCExtraBold",
            ShadowBlur = 0, ShadowColor = {0,0,0,0}, ShadowOffset={0, 3},
            Justification = "Right"
        })
    end

	wait(0.05)

	-- Clear Message
	RunClearMessagePresentation( screen, message )

	HandleScreenInput( screen )

end, RunControl )
